polkadot/node/network/collator-protocol/src/validator_side/collation.rs

```diff
@@ -267,33 +334,179 @@ impl Collations {
 		}
 	}
 
-	/// Checks the limit of seconded candidates.
+	/// Checks if another collation can be accepted. The number of collations that can be seconded
+	/// per parachain is limited by the entries in claim queue for the `ParaId` in question.
+	///
+	/// If prospective parachains mode is not enabled then we fall back to synchronous backing. In
+	/// this case there is a limit of 1 collation per relay parent.
+	///
+	/// If prospective parachains mode is enabled but claim queue is not supported then up to
+	/// `max_candidate_depth + 1` seconded collations are accepted. In theory in this case if two
+	/// parachains are sharing a core no fairness is guaranteed between them and the faster one can
+	/// starve the slower one by exhausting the limit with its own advertisements. In practice this
+	/// should not happen because core sharing implies core time support which implies the claim
+	/// queue being available.
 	pub(super) fn is_seconded_limit_reached(
 		&self,
 		relay_parent_mode: ProspectiveParachainsMode,
+		para_id: ParaId,
 	) -> bool {
-		let seconded_limit =
-			if let ProspectiveParachainsMode::Enabled { max_candidate_depth, .. } =
-				relay_parent_mode
+		let seconded_for_para = *self.seconded_per_para.get(&para_id).unwrap_or(&0);
+
+		match relay_parent_mode {
+			ProspectiveParachainsMode::Disabled => {
+				gum::trace!(
+					target: LOG_TARGET,
+					?para_id,
+					seconded_per_para=?self.seconded_per_para,
+					"is_seconded_limit_reached - ProspectiveParachainsMode::Disabled"
+				);
+
+				seconded_for_para >= 1
+			},
+			ProspectiveParachainsMode::Enabled { max_candidate_depth, allowed_ancestry_len: _ }
+				if !self.claim_queue_state.is_some() =>
 			{
-				max_candidate_depth + 1
-			} else {
-				1
-			};
-		self.seconded_count >= seconded_limit
+				gum::trace!(
+					target: LOG_TARGET,
+					?para_id,
+					seconded_per_para=?self.seconded_per_para,
+					max_candidate_depth,
+					"is_seconded_limit_reached - ProspectiveParachainsMode::Enabled without claim queue support"
+				);
+
+				seconded_for_para > max_candidate_depth
+			},
+			ProspectiveParachainsMode::Enabled {
+				max_candidate_depth: _,
+				allowed_ancestry_len: _,
+			} => {
+				let pending_for_para = self.pending_for_para(para_id);
+
+				let respected_per_para_limit =
+					self.claims_per_para.get(&para_id).copied().unwrap_or_default() >
+						seconded_for_para + pending_for_para;
+
+				gum::trace!(
+					target: LOG_TARGET,
+					?para_id,
+					claims_per_para=?self.claims_per_para,
+					seconded_per_para=?self.seconded_per_para,
+					?pending_for_para,
+					?respected_per_para_limit,
+					"is_seconded_limit_reached - ProspectiveParachainsMode::Enabled with claim queue support"
+				);
+
+				!respected_per_para_limit
+			},
+		}
+	}
+
+	/// Adds a new collation to the waiting queue for the relay parent. This function doesn't
+	/// perform any limits check. The caller (`enqueue_collation`) should assure that the collation
```

Comment:
Please don't assume who is the caller. It is fine to specify in the contract of the function that some invariant needs to be ensured by the caller, but don't assume who that is.

Reply:
I wanted to make the life the next person studying the code easier but I guess I overdid it. Fixed.